mobile/src/main/java/org/openhab/habdroid/ui/homescreenwidget/ItemUpdateWidget.kt

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.openhab.habdroid.ui.homescreenwidget
+
+import android.app.PendingIntent
+import android.appwidget.AppWidgetManager
+import android.appwidget.AppWidgetProvider
+import android.content.Context
+import android.content.Context.MODE_PRIVATE
+import android.content.Intent
+import android.graphics.BitmapFactory
+import android.net.Uri
+import android.os.Build
+import android.os.Bundle
+import android.util.Log
+import android.widget.RemoteViews
+import androidx.annotation.Px
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.openhab.habdroid.R
+import org.openhab.habdroid.background.BackgroundIntentReceiveActivity
+import org.openhab.habdroid.background.BackgroundTasksManager
+import org.openhab.habdroid.core.connection.ConnectionFactory
+import org.openhab.habdroid.ui.AbstractItemPickerActivity
+import org.openhab.habdroid.ui.PreferencesActivity
+import org.openhab.habdroid.util.HttpClient
+import org.openhab.habdroid.util.bitmapToSvg
+import org.openhab.habdroid.util.dpToPixel
+import org.openhab.habdroid.util.getIconFormat
+import org.openhab.habdroid.util.getPrefs
+import org.openhab.habdroid.util.getString
+import org.openhab.habdroid.util.isIconFormatPng
+import java.io.IOException
+import kotlin.math.min
+
+open class ItemUpdateWidget : AppWidgetProvider() {
+    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
+        Log.d(TAG, "onUpdate()")
+        appWidgetIds.forEach {
+            Log.d(TAG, "id: $it")
+            val data = getInfoForWidget(context, it)
+            if (data != null) {
+                setupWidget(context, data, it, appWidgetManager)
+            }
+        }
```

Comment:
I don't know if it's overkill, but I'd find this more elegant:
```
appWidgetIds.map { id -> Pair(id, getInfoForWidget(context, id)) }
    .filter { id, data -> data != null }
    .forEach { id, data -> setupWidget(context, data, id, appWidgetManager }

Reply:
By making the return type of `getInfoForWidget()` non-nullable, this can be simplified to:
````kotlin
        appWidgetIds.forEach {
            Log.d(TAG, "id: $it")
            setupWidget(context, getInfoForWidget(context, it), it, appWidgetManager)
        }
````